# Experiment 1


**Exploring Large Language Models in Programming Tasks**

In diesem Notebook untersuchen wir die Fähigkeiten verschiedener Large Language Models (LLMs) bei der Lösung von Programmieraufgaben. Die ausgewählten Modelle umfassen **Gemini**, **Deepseek**, **CodeLlama** sowie verschiedene Instanzen von **ChatGPT**. Unser Ziel ist es, herauszufinden, wie gut diese Modelle bei der Erstellung funktionaler Programme abschneiden, die spezifische Aufgaben in Python lösen.

Wir werden drei grundlegende Programmieraufgaben verwenden, um die Leistung jedes Modells zu testen:

1. **Ein Taschenrechner:** Wir testen, wie gut die Modelle grundlegende mathematische Operationen implementieren und wie robust ihre Fehlerbehandlung ist.
2. **Ein Einzelspieler PONG-Spiel:** Hier möchten wir überprüfen, ob die Modelle in der Lage sind, ein funktionierendes Spiel zu entwickeln und dabei sinnvolle Spielmechaniken zu integrieren.
3. **Sortieren von Zahlen:** Die Modelle werden aufgefordert, eine Liste von Zahlen zu sortieren. Dabei schauen wir uns an, welchen Algorithmus jedes Modell wählt und wie effizient dieser ist.

Im Laufe dieses Projekts werden wir uns auf die Korrektheit der Ausgaben, die verwendeten Bibliotheken, die Laufzeitleistung sowie die allgemeine Qualität des Codes konzentrieren.

Ziel des Projekts:
Das Hauptziel dieser Untersuchung ist es, zu verstehen, wie LLMs mit unterschiedlichen Programmieraufgaben umgehen und wie die generierten Lösungen im Vergleich zueinander abschneiden. Dies wird uns wertvolle Einblicke in die Stärken und Schwächen der einzelnen Modelle geben und uns helfen, die besten Modelle für verschiedene Anwendungsfälle zu identifizieren.


## Calculator

__Prompt__: Write a Python program for a basic calculator that supports addition, subtraction, multiplication, and division.

### CodeLlama 13b:

Folgende Antwort gibt CodeLlama 13b:

```python
class Calculator:
    def add(self, num1, num2):
        return num1 + num2
    
    def subtract(self, num1, num2):
        return num1 - num2
    
    def multiply(self, num1, num2):
        return num1 * num2
    
    def divide(self, num1, num2):
        if num2 == 0:
            raise ValueError("Cannot divide by zero")
        return num1 / num2
```
You can use this class as follows:
```python
calc = Calculator()
result = calc.add(3, 4)
print(result) # Output: 7

result = calc.subtract(3, 4)
print(result) # Output: -1

result = calc.multiply(3, 4)
print(result) # Output: 12

result = calc.divide(6, 2)
print(result) # Output: 3
```
You can also use the `+`, `-`, `*`, and `/` operators to perform the same operations:
```python
result = calc + 3 + 4
print(result) # Output: 7

result = calc - 3 - 4
print(result) # Output: -1

result = calc * 3 * 4
print(result) # Output: 12

result = calc / 6 / 2
print(result) # Output: 3
```
Note that the `divide` method raises a `ValueError` if the divisor is zero. You can catch this error and handle it as needed in your program.

CodeLlama gibt einmal die Calculator Klasse aus, sowie 2 Varianten, wie sich die Klasse nutzen lässt.

#### Code-Analyse

Die Klasse Calculator enthält vier grundlegende Methoden: add, subtract, multiply, und divide.
Jede Methode akzeptiert zwei numerische Eingabewerte (num1 und num2) und führt eine Operation basierend auf der jeweiligen Methode aus.
Diese Implementierung ermöglicht die grundlegenden arithmetischen Operationen, die man von einem Taschenrechner erwarten würde

In [5]:
class Calculator:
    def add(self, num1, num2):
        return num1 + num2
    
    def subtract(self, num1, num2):
        return num1 - num2
    
    def multiply(self, num1, num2):
        return num1 * num2
    
    def divide(self, num1, num2):
        if num2 == 0:
            raise ValueError("Cannot divide by zero")
        return num1 / num2

Erstes Anwendungsbeispiel instanziert das Calculator-Object und führt jede Operation einmal mit verschiedenen ganzan Zahlen durch. Beispiele funktionieren

In [6]:

calc = Calculator()
result = calc.add(3, 4)
print(result) # Output: 7

result = calc.subtract(3, 4)
print(result) # Output: -1

result = calc.multiply(3, 4)
print(result) # Output: 12

result = calc.divide(6, 2)
print(result) # Output: 3


7
-1
12
3.0


Der Code versucht, Operatoren wie +, -, *, und / direkt auf die Calculator-Instanz anzuwenden, was in Python nicht funktioniert. Die Calculator-Klasse hat keine Methoden zur Überladung dieser Operatoren. Ohne diese Überladung wirft Python einen TypeError, da calc nicht wie eine Zahl behandelt werden kann. Um dies zu ermöglichen, müsste die Klasse entsprechende Methoden wie __add__, __sub__ usw. implementieren.

In [7]:
result = calc + 3 + 4
print(result) # Output: 7

result = calc - 3 - 4
print(result) # Output: -1

result = calc * 3 * 4
print(result) # Output: 12

result = calc / 6 / 2
print(result) # Output: 3


TypeError: unsupported operand type(s) for +: 'Calculator' and 'int'

#### eigene Tests

##### 1. Division durch 0

In [9]:
calc = Calculator()
try:
    result = calc.divide(5, 0)
except ValueError as e:
    print(e)

Cannot divide by zero


##### 2. Extrem große Zahlen

In [13]:
result = calc.add(1e100, 1e100)
print(result)  # Ausgabe: 2e100

result = calc.subtract(1e100, 1e100)
print(result)  # Ausgabe: 0.0

result = calc.multiply(1e100, 1e100)
print(result)  # Ausgabe: 1e200

result = calc.divide(1e100, 1e100)
print(result)  # Ausgabe: 1.0


2e+100
0.0
1e+200
1.0


##### 3.Fließkommazahlen

In [12]:
result = calc.add(3.5, 4.2)
print(result)  # Ausgabe: 7.7

result = calc.subtract(3.5, 4.2)
print(result)  # Ausgabe: -0.7

result = calc.multiply(3.5, 4.2)
print(result)  # Ausgabe: 14.7

result = calc.divide(7.5, 2.5)
print(result)  # Ausgabe: 3.0


7.7
-0.7000000000000002
14.700000000000001
3.0


Die Werte -0.7000000000000002 und 14.700000000000001 entstehen durch die begrenzte Genauigkeit der Fließkommazahlen in der Computerarithmetik, bei der manche Dezimalzahlen nicht exakt dargestellt werden können, was zu minimalen Abweichungen führt.

##### 4. Nähe zu Null

In [14]:
result = calc.add(1e-10, 1e-10)
print(result)  # Ausgabe: 2e-10

result = calc.subtract(1e-10, 1e-10)
print(result)  # Ausgabe: 0.0

result = calc.multiply(1e-10, 1e10)
print(result)  # Ausgabe: 1.0

result = calc.divide(1e-10, 1e10)
print(result)  # Ausgabe: 1e-20

2e-10
0.0
1.0
1.0000000000000001e-20


##### 5.Negative Division

In [15]:
result = calc.divide(-6, 3)
print(result)  # Ausgabe: -2.0

result = calc.divide(6, -3)
print(result)  # Ausgabe: -2.0

result = calc.divide(-6, -3)
print(result)  # Ausgabe: 2.0


-2.0
-2.0
2.0


##### 6. Große Division (Vermeidung von Überlauffehlern)

In [17]:
result = calc.divide(1e100, 1e50)
print(result)  # Ausgabe: 1e50

9.999999999999999e+49


Rückgabe der unpräzisen Darstellung im wissenschaftlichen Format

##### 7. Falsche Argumente

In [20]:
result = calc.add("abc", 5)
print("Expected TypeError:", result)

TypeError: can only concatenate str (not "int") to str

TypeError wird nicht abgefangen